File: Submissions/Limit Breakers/E-Commerce Store Scenario/src/CustomerDB.java
class CustomerDB {
    String customerId;
    String name;
    int purchaseCount; // used for promotion heuristic
    String email;

    public CustomerDB(String customerId, String name, String email) {
        this.customerId = customerId;
        this.name = name;
        this.email = email;
        this.purchaseCount = 0;
    }

    public void incrementPurchases() {
        purchaseCount++;
    }

    public String toString() {
        return "CustomerID=" + customerId + ", Name=" + name + ", Purchases=" + purchaseCount;
    }
}

class CustomerNode {
    CustomerDB data;
    CustomerNode left, right;

    public CustomerNode(CustomerDB data) {
        this.data = data;
    }
}

class CustomerBST {
    private CustomerNode root;

    public void addCustomer(CustomerDB profile) {
        root = insert(root, profile.customerId, profile);
    }

    private CustomerNode insert(CustomerNode node, String id, CustomerDB profile) {
        if (node == null) return new CustomerNode(profile);
        if (id.compareTo(node.data.customerId) < 0)
            node.left = insert(node.left, id, profile);
        else if (id.compareTo(node.data.customerId) > 0)
            node.right = insert(node.right, id, profile);
        return node;
    }

    public CustomerDB findCustomer(String id) {
        CustomerNode node = find(root, id);
        return (node != null) ? node.data : null;
    }

    private CustomerNode find(CustomerNode node, String id) {
        if (node == null || node.data.customerId.equals(id)) return node;
        if (id.compareTo(node.data.customerId) < 0) return find(node.left, id);
        else return find(node.right, id);
    }

    public void deleteCustomer(String id) {
        root = delete(root, id);
    }

    private CustomerNode delete(CustomerNode node, String id) {
        if (node == null) return null;

        if (id.compareTo(node.data.customerId) < 0)
            node.left = delete(node.left, id);
        else if (id.compareTo(node.data.customerId) > 0)
            node.right = delete(node.right, id);
        else {
            if (node.left == null) return node.right;
            if (node.right == null) return node.left;
            CustomerNode min = findMin(node.right);
            node.data = min.data;
            node.right = delete(node.right, min.data.customerId);
        }
        return node;
    }

    private CustomerNode findMin(CustomerNode node) {
        while (node.left != null) node = node.left;
        return node;
    }

    public void promoteHotCustomer(String id) {
        CustomerNode parent = null;
        CustomerNode node = root;

        while (node != null && !node.data.customerId.equals(id)) {
            parent = node;
            if (id.compareTo(node.data.customerId) < 0) node = node.left;
            else node = node.right;
        }

        if (node == null) {
            System.out.println("Customer not found for promotion.");
            return;
        }

        if (parent != null && parent.left == node) {
            parent.left = node.right;
            node.right = parent;

            if (root == parent) root = node;
            System.out.println("Promoted hot customer " + id + " closer to root.");
        } else if (parent != null && parent.right == node) {
            parent.right = node.left;
            node.left = parent;
            if (root == parent) root = node;
            System.out.println("Promoted hot customer " + id + " closer to root.");
        } else {
            System.out.println("Customer already at root.");
        }
    }

    public void display() {
        System.out.println("\nCustomer BST (Inorder):");
        inorder(root);
    }

    private void inorder(CustomerNode node) {
        if (node == null) return;
        inorder(node.left);
        System.out.println("  " + node.data);
        inorder(node.right);
    }
}

File: Submissions/CompileHer/E-Commerce Store Scenario/datastructures/customer_db.py
class CustomerNode:
    def __init__(self, id, profile):
        self.id = id
        self.profile = profile
        self.left = None
        self.right = None
        self.parent = None 
        self.purchase_frequency = profile.get('freq', 0)
        self.rec_graph = profile.get('rec_graph', {}) 

    def __repr__(self):
        return f"CustNode(ID: {self.id}, Freq: {self.purchase_frequency})"

class CustomerDB:
    def __init__(self):
        self.root = None
      

    def addCustomer(self, id, profile):
        new_node = CustomerNode(id, profile)
        if not self.root:
            self.root = new_node
            return new_node
        
        current = self.root
        while True:
            if id < current.id:
                if current.left is None:
                    current.left = new_node
                    new_node.parent = current
                    return new_node
                current = current.left
            else: 
                if current.right is None:
                    current.right = new_node
                    new_node.parent = current
                    return new_node
                current = current.right

    def findCustomer(self, id):
        current = self.root
        while current:
            if id == current.id:
                return current
            elif id < current.id:
                current = current.left
            else:
                current = current.right
        return None

    def deleteCustomer(self, id):
        return self.findCustomer(id) is not None


    def _rotate_right(self, y_node):
        x_node = y_node.left
        y_node.left = x_node.right
        if x_node.right:
            x_node.right.parent = y_node
        
        x_node.parent = y_node.parent
        if not y_node.parent:
            self.root = x_node
        elif y_node == y_node.parent.right:
            y_node.parent.right = x_node
        else:
            y_node.parent.left = x_node
            
        x_node.right = y_node
        y_node.parent = x_node
        return x_node

    def _rotate_left(self, x_node):
        y_node = x_node.right
        x_node.right = y_node.left
        if y_node.left:
            y_node.left.parent = x_node
        
        y_node.parent = x_node.parent
        if not x_node.parent:
            self.root = y_node
        elif x_node == x_node.parent.left:
            x_node.parent.left = y_node
        else:
            x_node.parent.right = y_node
            
        y_node.left = x_node
        x_node.parent = y_node
        return y_node


    def promoteHotCustomer(self, id):

        node = self.findCustomer(id)
        if not node or not node.parent:
            return {"status": "Customer not found or is already the root."}
        
        parent = node.parent
        
        if node == parent.left:
            self._rotate_right(parent)
            return {"status": f"Customer {id} promoted one level (Right Rotation)."}
        
        elif node == parent.right:
            self._rotate_left(parent)
            return {"status": f"Customer {id} promoted one level (Left Rotation)."}
        
        return {"status": "Promotion logic failed."}
    

    def recommendProducts(self, customer_id, k):

        node = self.findCustomer(customer_id)
        if not node:
            return {"status": "Customer not found."}
        
        recommendation_counts = {}
        
        for product_id in node.rec_graph:
            for co_bought_id, count in node.rec_graph[product_id].items():
                
                if co_bought_id not in node.rec_graph:
                    recommendation_counts[co_bought_id] = recommendation_counts.get(co_bought_id, 0) + count

        sorted_recs = sorted(
            recommendation_counts.items(), 
            key=lambda item: item[1], 
            reverse=True
        )

        top_k_recs = [item[0] for item in sorted_recs[:k]]
        
        return {"status": "Recommendations generated.", "recommendations": top_k_recs}
